blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,249 @@
+package blockbuilder
+
+import (
+	"github.com/0xPolygon/polygon-edge/consensus"
+	"github.com/0xPolygon/polygon-edge/consensus/ibft/signer"
+	"github.com/0xPolygon/polygon-edge/helper/common"
+	"github.com/0xPolygon/polygon-edge/state"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/hashicorp/go-hclog"
+)
+
+// TODO: Add open tracing
+
+const (
+	BlockGasTargetDivisor uint64 = 1024 // The bound divisor of the gas limit, used in update calculations
+)
+
+// BlockBuilderParams are fields for the block that cannot be changed
+type BlockBuilderParams struct {
+	// Parent block
+	Parent *types.Header
+
+	// Executor. TODO: Abstract executor
+	Executor *state.Executor
+
+	// Coinbase that is signing the block
+	Coinbase types.Address
+
+	// Vanity extra for the block
+	VanityExtra []byte
+
+	// GasLimit is the gas limit for the block
+	BlockGasTarget uint64
+
+	// Logger
+	Logger hclog.Logger
+
+	// txPoolInterface implementation
+	TxPool txPoolInterface
+}
+
+type blockchainInterface interface {
+	// WriteBlock writes the block to the db but does not process it
+	WriteBlock(b *types.Block, source string) error
+}
+
+type txPoolInterface interface {
+	Prepare()
+	Peek() *types.Transaction
+	Pop(tx *types.Transaction)
+	Drop(tx *types.Transaction)
+	Demote(tx *types.Transaction)
+}
+
+func NewBlockBuilder(params *BlockBuilderParams) (*BlockBuilder, error) {
+	// vanity extra can only be 32 size max. it is better to trim that to return
+	// an error that we have to propagate. It should be up to higher level
+	// code to error if the vanity extra supplied by the user is too big.
+	if len(params.VanityExtra) > 32 {
+		params.VanityExtra = params.VanityExtra[:32]
+	}
+
+	if params.VanityExtra == nil {
+		params.VanityExtra = make([]byte, 0)
+	}
+
+	if params.Logger == nil {
+		params.Logger = hclog.NewNullLogger()
+	} else {
+		params.Logger = params.Logger.Named("block-builder")
+	}
+
+	builder := &BlockBuilder{
+		params: params,
+	}
+	if err := builder.Reset(); err != nil {
+		return nil, err
+	}
+
+	return builder, nil
+}
+
+type BlockBuilder struct {
+	// input params for the block
+	params *BlockBuilderParams
+
+	// header is the header for the block being created
+	header *types.Header
+
+	// transactions are the data included in the block
+	txns []*types.Transaction
+
+	// block is a reference to the already built block
+	block *types.Block
+
+	state *state.Transition
+}
+
+func (b *BlockBuilder) Reset() error {
```

Comment:
`Reset` seems not to be called after BlockBuilder is used. Right now `Init` is proper?

Reply:
We use this functionality in the `v3-parity` branch to create a single instance of `block-builder` and reset it during rounds. With this, we manage to compute the block only once during sealing. I would keep it since we can streamline the review later on of that branch.

Reply:
Noted. Will resolve